server/server.go

```diff
@@ -3,52 +3,162 @@ package server
 import (
 	"archive/tar"
 	"archive/zip"
+	"bytes"
 	"coderunner/constants"
+	"compress/gzip"
 	"encoding/json"
 	"fmt"
 	"io"
 	"log"
 	"mime/multipart"
 	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
 	"sync"
 )
 
-// Parses the client request and uploads the file.
-func upload(w http.ResponseWriter, r *http.Request) {
-	fmt.Println("File Upload Endpoint Hit")
+type assignmentTestingInformation struct {
+	CommandToExecute string
+	CommandToCompile string
+	WorkDir          string
+	RootDir          string
+	Output           string
+	CmdlineArgs      map[string]string
+}
 
+var assignTestingInfo assignmentTestingInformation
+
+// upload parses the client request and uploads the file.
+func upload(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Access-Control-Allow-Origin", "*")
 
-	responseString := readFormData(r)
+	output := readFormData(r)
 
-	if len(responseString) <= 0 {
-		responseString += `"Upload Status":"Successfully Uploaded File(s)"`
+	if len(output) <= 0 {
+		output += `"Upload Status":"Successfully Uploaded File(s)"`
 	}
 
-	responseS, err := json.Marshal(responseString)
+	response, err := json.Marshal(output)
 	if err != nil {
 		log.Println(err)
 	}
 
 	// Write the response to be sent to client.
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
-	w.Write(responseS)
+	w.Write(response)
 }
 
-// Reads the compressed file and invokes function to decompress it
+// buildRun builds and runs the assignment uploaded.
+func buildRun(w http.ResponseWriter, r *http.Request) {
+
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	var outputString string
+	var currDir string
+	var err error
+
+	// Navigate to the assignment working directory.
+	outputString, currDir = navigateToWorkDir()
+	if outputString == "" {
+		// Execute the compile command.
+		outputString, err = runCommand(assignTestingInfo.CommandToCompile)
+
+		if err == nil {
+			// Append the command line arguments to run command.
+			runCmd := assignTestingInfo.CommandToExecute
+			for _, value := range assignTestingInfo.CmdlineArgs {
+				runCmd = fmt.Sprintf("%s %s", runCmd, value)
+			}
+			// Execute the assignment run command.
+			outputString, err = runCommand(runCmd)
+			if err != nil {
+				log.Println("error while executing the assignment", err)
+				outputString = err.Error()
+			}
+		}
+
+		// Navigate back to the code-runner working directory after successful execution.
+		err = os.Chdir(currDir)
+		if err != nil {
+			log.Println("error while navigating to the current directory", err)
+		}
+	}
+
+	response, err := json.Marshal(outputString)
+	if err != nil {
+		log.Println(err)
+	}
+	// Write the response to be sent to client.
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
```

Comment:
raise an issue to return semantic http status codes back to the client.

Reply:
ok